c/tree.c

```diff
@@ -629,37 +640,190 @@ LOCAL MMDBW_status insert_record_for_network(
+
     if (merge_strategy == MMDBW_MERGE_STRATEGY_UNKNOWN) {
         merge_strategy = tree->merge_strategy;
     }
-    MMDBW_record_s *record_to_set, *other_record, *parent_record;
-    MMDBW_status status =
-        find_record_for_network(tree, network,
-                                false,
-                                &new_node_from_record,
-                                &record_to_set, &other_record,
-                                &parent_record);
-    if (MMDBW_SUCCESS != status) {
-        status = free_record_value(tree, new_record, true);
+
+    return insert_record_into_next_node(
+               tree,
+               &(tree->root_record),
+               network,
+               0,
+               new_record,
+               merge_strategy,
+               is_internal_insert
+               );
+}
+
+LOCAL MMDBW_status insert_record_into_next_node(
+    MMDBW_tree_s *tree,
+    MMDBW_record_s *current_record,
+    MMDBW_network_s *network,
+    int current_bit,
+    MMDBW_record_s *new_record,
+    MMDBW_merge_strategy merge_strategy,
+    bool is_internal_insert)
+{
+    if (current_bit >= network->prefix_length
+        && (current_record->type == MMDBW_RECORD_TYPE_EMPTY
+            || current_record->type == MMDBW_RECORD_TYPE_DATA
+            || ( current_record->type == MMDBW_RECORD_TYPE_NODE
+                 && merge_strategy == MMDBW_MERGE_STRATEGY_NONE))) {
+        return insert_record_into_current_record(
+                   tree,
+                   current_record,
+                   network,
+                   new_record,
+                   merge_strategy,
+                   is_internal_insert
+                   );
+    }
+
+    MMDBW_node_s *next_node = NULL;
+    switch (current_record->type) {
+    case MMDBW_RECORD_TYPE_EMPTY:
+    case MMDBW_RECORD_TYPE_DATA: {
+            next_node = new_node_from_record(tree, current_record);
+            current_record->value.node = next_node;
+            current_record->type = MMDBW_RECORD_TYPE_NODE;
+            break;
+        }
+    case MMDBW_RECORD_TYPE_ALIAS: {
+            // The insert is trying to overwrite an aliased network. We do not
+            // allow this.
+            if (current_bit == network->prefix_length) {
+                return MMDBW_ALIAS_OVERWRITE_ATTEMPT_ERROR;
+            }
+            // We don't follow aliases when inserting a network that contains an
+            // aliased network within it.
+            else if (current_bit > network->prefix_length) {
+                return MMDBW_SUCCESS;
+            } else {
+                return MMDBW_INSERT_INTO_ALIAS_NODE_ERROR;
+            }
+        }
+    case MMDBW_RECORD_TYPE_FIXED_NODE:
+    case MMDBW_RECORD_TYPE_NODE: {
+            next_node = current_record->value.node;
+            break;
+        }
+    }
+
+    // If we are inserting a fixed node or an alias, we make all of the nodes
+    // down to that record fixed nodes. This makes it easier to not accidently
```

Comment:
If you are referring to "accidently", [Webster disagrees](https://www.merriam-webster.com/dictionary/accidental). That said, I fixed it.